src/session.rs

```diff
@@ -0,0 +1,137 @@
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::fs;
+use std::path::PathBuf;
+
+use anyhow::Result;
+use cookie::Cookie;
+use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, COOKIE, SET_COOKIE};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Content {
+    #[serde(rename = "__meta__")]
+    meta: HashMap<String, String>,
+    // TODO: Replace String with Cookie
+    // (need to use Serde's serialize_with and deserialize_with)
+    cookies: HashMap<String, String>,
+    headers: HashMap<String, String>,
+}
+
+impl Default for Content {
+    fn default() -> Self {
+        let mut meta = HashMap::new();
+        meta.insert("about".into(), "xh session file".into());
+        meta.insert("xh".into(), env!("CARGO_PKG_VERSION").into());
+        Content {
+            meta,
+            cookies: HashMap::new(),
+            headers: HashMap::new(),
+        }
+    }
+}
+
+pub struct Session {
+    pub path: PathBuf,
+    pub read_only: bool,
+    content: Content,
+}
+
+fn ensure_file_exists(path: PathBuf) -> Result<PathBuf> {
+    if path.exists() {
+        Ok(path)
+    } else {
+        fs::create_dir_all(path.parent().unwrap())?;
+        fs::File::create(&path)?;
+        Ok(path)
+    }
+}
+
+impl Session {
+    pub fn load_session(host: &str, name_or_path: &str, read_only: bool) -> Result<Self> {
```

Comment:
I think HTTPie's behavior might be a bit better. As far as I can tell:
1. If there's no file, it doesn't create it until it successfully completed the request.
2. An empty file is rejected. A session file should always have valid JSON in it, if it exists.
3. If it didn't exist yet, a file is always created after a successful request, even if `--session-read-only` is used.

Doing it like that means we don't have to handle empty files, and there's no risk of leaving empty files behind. I think it's best not to modify the filesystem at all in `load_session()`.

So the new way would be to try to open the file, and
- If it's not found, make a default session and only try to create it when persisting.
- If there's some other error, bubble it up.
- If decoding fails, bubble it up.